### DISCUSSION TITLE
Approximation of K-means clustering?

### DISCUSSION THREAD

**Discussion by kipcole9** [libvips/libvips#2840] [Help wanted]
Any advice on a libvips mechanism to do k-means clustering on an image?  Basically to identify the most common colours in an image, and the number of pixels that are in that cluster. Similar to Method 3 described [here](https://towardsdatascience.com/finding-most-common-colors-in-python-47ea0767a06a) but using libvips only.

**Comment by kipcole9** [comment]
Perhaps a public API version of [vips__quantise_get_palette](https://github.com/libvips/libvips/blob/6aca41d7135be9d14cbc04f406958f603698bade/libvips/foreign/quantise.c#L97-L101) since that already wraps around [libimagequant](https://github.com/ImageOptim/libimagequant)'s [liq_get_pallete](https://pngquant.org/lib/#liq_get_palette) function?

**Comment by jcupitt** [comment]
Sure, good idea. libvips has N-dimensional histograms (something like method 2 I guess?) but k-mean would be better in many cases.

We're finishing up 8.13 now, so let's put this on the list for 8.14.

**Comment by jcupitt** [comment]
There's a PR now: https://github.com/libvips/libvips/pull/3122
